README.md

[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat)](http://bioconda.github.io/recipes/fusion-report/README.html)
[![CircleCI Build Status](https://circleci.com/gh/circleci/circleci-docs.svg?style=shield)](https://circleci.com/gh/BU-ISCIII/plasmidID) [![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0) [![Scif](https://img.shields.io/badge/Filesystem-Scientific-brightgreen.svg)](https://sci-f.github.io)

# plasmidID <img align="left" src="https://github.com/BU-ISCIII/plasmidID/blob/develop/img/plasmidID_logo.png" alt="Logo" width="100">

<br>
<br>

* [Introduction](#introduction)
* [Requirements](#requirements)
    * [Software](#software)
    * [Plasmid database](#plasmid-database)
* [Installation](#installation)
    * [Install from source](#install-from-source)
    * [Install using conda](#install-using-conda)
* [Quick usage](#quick-usage)
* [Usage](#usage)
* [Output](#output)
* [Annotation file](#annotation-file)
* [Illustrated pipeline](#illustrated-pipeline)
* [Docker](#docker)

## Introduction

PlasmidID is a mapping-based, assembly-assisted plasmid identification tool that analyzes and gives graphic solution for plasmid identification.

PlasmidID is a **computational pipeline** implemented in **BASH** that maps Illumina reads over plasmid database sequences. The k-mer filtered, most covered sequences are clustered by identity to avoid redundancy and the longest are used as scaffold for plasmid reconstruction. Reads are assembled and annotated by automatic and specific annotation. All information generated from mapping, assembly, annotation and local alignment analyses is gathered and accurately represented in a **circular image** which allow user to determine plasmidic composition in any bacterial sample.

## Requirements

#### Software

* [Python >=3.6](https://www.python.org/)
* [Trimmomatic v0.33](http://www.usadellab.org/cms/?page=trimmomatic)(Optional)
* [Spades v3.8](http://cab.spbu.ru/software/spades/) (Optional)
* [Perl v5.26.0](https://www.perl.org/get.html)
* [NCBI_blast + v2.2.3](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download)
* [Bedtools v2.25](http://bedtools.readthedocs.io/en/latest/)
* [Bowtie 2 v2.2.4](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
* [SAMtools v1.2](http://samtools.sourceforge.net/)
* [prokka v1.12](http://www.vicbioinformatics.com/software.prokka.shtml)
* [cd-hit v4.6.6](http://weizhongli-lab.org/cd-hit/) (no longer needed since v1.6)
* [circos v0.69.3](http://circos.ca/software/download/circos/)
* [mash v2.2](https://github.com/marbl/Mash)

#### Plasmid database

Since version v1.5.1 plasmid database can be downloaded with the following command:

```Bash
 download_plasmid_database.py -o FOLDER
```

## Installation

#### Install from source

Install all dependencies and add them to $PATH

git clone https://github.com/BU-ISCIII/plasmidID.git

Add plasmidID and ./bin to $PATH

#### Install using conda

This option is recomended.

Install [Anaconda3](https://www.anaconda.com/distribution/)

```
conda install -c conda-forge -c bioconda plasmidid
```
Wait for the environment to solve

Ignore warnings/errors

#### Use Docker

Example:
Clone the repo:
```Bash
git clone git@github.com:BU-ISCIII/plasmidID.git
cd plasmidID
```
Run it with the test data using docker:

**Notice that the input files MUST be in your present working directory or in any folder inside it. For example, if I execute this command in `/home/smonzon`, my folder with the files would be in `/home/smonzon/test`.**

```Bash
docker run -v $PWD:$PWD -w $PWD buisciii/plasmidid plasmidID \
     -1 test/KPN_TEST_R1.fastq.gz  \
     -2 test/KPN_TEST_R2.fastq.gz \
     -d test/plasmids_TEST_database.fasta \
     -c test/contigs_KPN_TEST.fasta \
     --no-trim \
     -s KPN
```

## Quick usage

Illumina paired-end
```
plasmidID \
-1 SAMPLE_R1.fastq.gz  \
-2 SAMPLE_R2.fastq.gz \
-d YYYY-MM-DD_plasmids.fasta \
-c SAMPLE_assembled_contigs.fasta \
--no-trim \
-s SAMPLE
```

SMRT sequencing (only contigs)
```
plasmidID \
-d YYYY-MM-DD_plasmids.fasta \
-c SAMPLE_contigs.fasta \
-s SAMPLE
```

Annotate any fasta you want
```
plasmidID \
-d YYYY-MM-DD_plasmids.fasta \
-c SAMPLE_assembled_contigs.fasta \
-a annotation_file \
-s SAMPLE
```
More info about [annotation file](#annotation-file)

If there are several samples in the same GROUP folder
```
summary_report_pid.py -i NO_GROUP/
```
## Usage

```
usage : plasmidID <-1 R1> <-2 R2> <-d database(fasta)> <-s sample_name> [-g group_name] [options]

	Mandatory input data:
	-1 | --R1	<filename>	reads corresponding to paired-end R1 (mandatory)
	-2 | --R2	<filename>	reads corresponding to paired-end R2 (mandatory)
	-d | --database	<filename>	database to map and reconstruct (mandatory)
	-s | --sample	<string>	sample name (mandatory), less than 37 characters

	Optional input data:
	-g | --group	<string>	group name (optional). If unset, samples will be gathered in NO_GROUP group
	-c | --contigs	<filename>	file with contigs. If supplied, plasmidID will not assembly reads
	-a | --annotate <filename>	file with configuration file for specific annotation
	-o 		<output_dir>	output directory, by default is the current directory

	Pipeline options:
	--explore	Relaxes default parameters to find less reliable relationships within data supplied and database
	--only-reconstruct	Database supplied will not be filtered and all sequences will be used as scaffold
						This option does not require R1 and R2, instead a contig file can be supplied
	-w 			Undo winner takes it all algorithm when clustering by kmer - QUICKER MODE
	Trimming:
	--trimmomatic-directory Indicate directory holding trimmomatic .jar executable
	--no-trim	Reads supplied will not be quality trimmed

	Coverage and Clustering:
	-C | --coverage-cutoff	<int>	minimun coverage percentage to select a plasmid as scafold (0-100), default 80
	-S | --coverage-summary	<int>	minimun coverage percentage to include plasmids in summary image (0-100), default 90
	-f | --cluster	<int>	kmer identity to cluster plasmids into the same representative sequence (0 means identical) 		(0-1), default 0.5
	-k | --kmer	<int>	identity to filter plasmids from the database with kmer approach (0-1), default 0.95

	Contig local alignment
	-i | --alignment-identity <int>	minimun identity percentage aligned for a contig to annotate, default 90
	-l | --alignment-percentage <int>	minimun length percentage aligned for a contig to annotate, default 20
	-L | --length-total	<int>	minimun alignment length to filter blast analysis
	--extend-annotation <int>	look for annotation over regions with no homology found (base pairs), default 500bp

	Draw images:
	--config-directory <dir>	directory holding config files, default config_files/
	--config-file-individual <file-name> file name of the individual file used to reconstruct
	Additional options:

	-M | --memory	<int>	max memory allowed to use
	-T | --threads	<int>	number of threads
	-v | --version		version
	-h | --help		display usage message

example: ./plasmidID.sh -1 ecoli_R1.fastq.gz -2 ecoli_R2.fastq.gz -d database.fasta -s ECO_553 -G ENTERO
	./plasmidID.sh -1 ecoli_R1.fastq.gz -2 ecoli_R2.fastq.gz -d PacBio_sample.fasta -c scaffolds.fasta -C 60 -s ECO_60 -G ENTERO --no-trim
```

## Examples

Under construction

## Output

Since v1.6, the more relevant output is located in GROUP/SAMPLE folder:

- **SAMPLE_final_results.html(.tab)**
	- id: Name of the accession number of reference
	- length: length of the reference sequence
	- species: species of the reference sequence
	- description: rest of reference fasta header
	- contig_name: number of the contigs that align the minimun required for complete contig track
	- SAMPLE:
		- Image of the reconstructed plasmid (click to open in new tab)
		- MAPPING % (percentage): percentage of reference covered with reads
			- X for contig mode (gray colour)
			- Orientative colouring (the closer to 100% the better)
		- ALIGN FR (fraction_covered): total length of contigs aligned (complete) / reference sequence length
			- Orientative colouring (the closer to 1 the better)


## Annotation file

Under construction

## Illustrated pipeline

This image sumarizes PlasmidID pipeline, including the most important steps.
For furder details, including:
- [Results interpretation](https://github.com/BU-ISCIII/plasmidID/wiki/Understanding-the-image:-track-by-track)
- and more, please visit: [**PLASMIDID WIKI**](https://github.com/BU-ISCIII/plasmidID/wiki)

<p align="center"><img src="https://github.com/BU-ISCIII/plasmidID/blob/master/img/pipeline_pID.png" alt="workflow_small"  width="500">



CHANGELOG.md
# Changelog

## 1.6.4 - 2021-03-2020
### Added
- Updated Dockerfile
- Migrated tests to github actions
### Fixed
- Updated environment.yml for conda.
- Fixed issues #12,#14,#15,#17. Cases with no plasmids or too many. Relative paths in html images.

## 1.4.2 - 2018-09-29
### Added
- Specific config file for only reconstruct parameter

###Fixed
- Protein databases can be properly used

## 1.4 - 2018-09-20
### Added
- Automatically annotated genes/cds are displayed differently depending on whether they are located in forward or reverse
- Psi-cd-hit and blast now handle threads
- Improved error handling
- Doocker/Singularity compatibility
- One multifasta file per reference plasmid is generated with all the similar contigs from the sample
- Quick staus of values applied to plasmid reconstruction

###Fixed
- Some plasmids from the database were not annotated
- Limit sample name to 37 characters, capped by prokka
- Bug in complete contig track generator that took the wrong value and couldn't draw sequences that matched the position 0 of plasmid



## 1.3.0 - 2018-07-11
### New
- Summary table can be generated with new utility
- Several databases can be now annotated filling annotation_config_file.txt
- --only-reconstruct is now implemented if user only needs to reconstruct and annotate contigs with small known databases
### Fixed
- circos dependency is now checked
- Output is now correctly redirected with -o
### Added
- trimmomatic directory containing .jar can no be especified with --trimmomatic-directory
- Vervose mode included. By default a log file will be created
- Friendly terminal output

## 1.2.2 - 2018-06-22
### Fixed
- ***IMPORTANT***: PlasmidID maps with -a mode NOW, as it should have allways been. A bug on mapping script is now solved
- Number of threads are now implemented on mapping
- Some cumulative clustering temporary files are now removed

## 1.2.1 - 2018-06-14
### Fixed
- All dependencies are now checked at the beggining
- Path to scripts are no longer hard coded paths
- Links should be now displayed on summary image

### Added
- Added first utility ***ncbi_database_fetcher.sh***, a script to download FASTA databases from terms
- Short scripts now moved to /bin has to be added to PATH


## 1.1.1 - 2018-06-11
### Fixed
- Additional database will not be required for circos executios, even though the file will be created
- Fixed an issue when no plasmid matches mapping requeriments
- Fixed an issue when circos will trow an error message when no plasmids met mapping requeriments


## 1.1.0 - 2018-06-06
### Added
- Database plasmids used as scaffold are annotated after filtering. User doesn't need to annotate the initial huge plasmid database.
- User can add ONE nucleotide FASTA file wi that will be specifically annotated on final plasmids with a light blue color

## Unreleased

- Create config files as required by user and include visual parameters
- Test and adapt the --only-reconstruct option


lib/cdhit_cluster.sh
#!/bin/bash

set -e

#=============================================================
# HEADER
#=============================================================

#INSTITUTION:ISCIII
#CENTRE:BU-ISCIII
#AUTHOR: Pedro J. Sola
VERSION=1.0 
#CREATED: 6 April 2018
#REVISION:
#DESCRIPTION:Script that uses cd-hit/psi-cd-hit to clusterize a FASTA file
#
#DOCUMENTATION
#
#
#Compare floats in BASH
#
#if [ $(echo "$cluster_cutoff > 0.7"|bc -l) -eq 1 ]; then
#	echo "YES"
#else
#	echo "NO"
#fi
#
#-d length of description in .clstr file, default 20. if set to 0, 
#	it takes the fasta defline and stops at first space
#-s length difference cutoff, default 0.0
#	if set to 0.9, the shorter sequences need to be
#	at least 90% length of the representative of the cluster
#-B 1 or 0, default 0, by default, sequences are stored in RAM
#	if set to 1, sequence are stored on hard drive
#	it is recommended to use -B 1 for huge databases
#-g 1 or 0, default 0
#	By cd-hit’s default algorithm, a sequence is clustered to the first
#	cluster that meet the threshold (fast mode). If set to 1, the program
#	will cluster it into the most similar cluster that meet the threshold
#	(accurate but slow mode)
#
#	PSI-CD-HIT
#-G (1/0) use global identity? default 1, sequence identity
#	calculated as total identical residues of local alignments
#	length of shorter sequence
#
#-n 5 for thresholds 0.7 ~ 1.0
#-n 4 for thresholds 0.6 ~ 0.7
#-n 3 for thresholds 0.5 ~ 0.6
#-n 2 for thresholds 0.4 ~ 0.5

#================================================================
# END_OF_HEADER
#================================================================

#SHORT USAGE RULES
#LONG USAGE FUNCTION
usage() {
	cat << EOF

Cdhit_cluster script uses cd-hit/psi-cd-hit to clusterize a FASTA file

usage : $0 <-i inputfile(FASTA)> [-o <directory>] [-n <filename>] [-c <percentage>]
		[-T <threads>] [-g group_name] [-s <int>] [-M <int>][-C <(0|1)>] [-G <(0|1)>] [-b <blast_prog>] [p] [-v] [-h]

	-i input file in FASTA format
	-c percentage threshold to cluster, default 80
	-M max available memory (Mbyte), default 400
	-n file name
	-s length difference cutoff, default 0.8
	-g group name (optional). If unset, samples will be gathered in NO_GROUP group
	-p runs psi-cd-hit instead of cd-hit
	-C psi-cd-hit only: circular sequences, default 1. If set to 0 sequence is assumed lineal
	-G psi-cd-hit only: gobal identity, -G 0 only takes the first local alignment for clustering
	-b psi-cd-hit only: choose blast program, default blastn
	-T number of threads
	-v version
	-h display usage message


Output directory is the same as input directory

example: cdhit_cluster -i ecoli.fasta -c 90 -M 50000 -T 0
		 

EOF
}

#================================================================
# OPTION_PROCESSING
#================================================================
#Make sure the script is executed with arguments
if [ $# = 0 ] ; then
 usage >&2
 exit 1
fi


#DECLARE FLAGS AND VARIABLES
cwd="$(pwd)"
group="NO_GROUP"
input_file="Input_file"
cluster_cutoff=0.8
max_memory=400
length_cutoff=0.8
cd_hit_command=cd-hit-est
is_circle=1
global_psi_cd_hit=1
psi_cd_hit_program=blastn
word_size=0
threads=0

#PARSE VARIABLE ARGUMENTS WITH getops
#common example with letters, for long options check longopts2getopts.sh
options=":i:o:c:M:n:s:g:C:G:b:T:pvh"
while getopts $options opt; do
	case $opt in
		i )
			input_file=$OPTARG
			;;
		
		c )
			cluster_cutoff_input=$OPTARG
			;;
		g)
			group=$OPTARG
			;;
		M )
			max_memory=$OPTARG
			;;
		n )
			file_name=$OPTARG
			;;
		s )			
          	length_cutoff=$OPTARG
          	;;
        p )			
          	cd_hit_command=psi-cd-hit.pl
          	;;
        C )			
          	is_circle=$OPTARG
          	;;
        G)			
          	global_psi_cd_hit=$OPTARG
          	;;
        T)			
          	threads=$OPTARG
          	;;
        b)			
          	psi_cd_hit_program=$OPTARG
          	;;  	
        h )
		  	usage
		  	exit 1
		  	;;
		v )
		  	echo $VERSION
		  	exit 1
		  	;;
		\?)  
			echo "Invalid Option: -$OPTARG" 1>&2
			usage
			exit 1
			;;
		: )
      		echo "Option -$OPTARG requires an argument." >&2
      		exit 1
      		;;
      	* ) 
			echo "Unimplemented option: -$OPTARG" >&2;
			exit 1
			;;

	esac
done
shift $((OPTIND-1))

#================================================================
# MAIN_BODY
#================================================================
##CHECK DEPENDENCIES, MANDATORY FIELDS, FOLDERS AND ARGUMENTS

echo -e "\n#Executing" $0 "\n"

bash lib/check_mandatory_files.sh $input_file

bash lib/check_dependencies.sh cd-hit-est psi-cd-hit.pl



# Set word size (parameter -n for cd-hit) as author recomends
#according to clustering percentage


cluster_cutoff=$(echo "$cluster_cutoff_input / 100" | bc -l | sed 's/0\{1,\}$//')
#cluster_cutoff=${cluster_cutoff%.*} #Remove float value


if [[ "$cluster_cutoff_input" -gt 70  &&  "$cluster_cutoff_input" -le 100 ]]; then
	word_size=5
elif [[ "$cluster_cutoff_input" -gt 60  &&  "$cluster_cutoff_input" -le 70 ]]; then
	word_size=4
elif [[ "$cluster_cutoff_input" -gt 50  &&  "$cluster_cutoff_input" -le 60 ]]; then
	word_size=3
elif [[ "$cluster_cutoff_input" -ge 40  &&  "$cluster_cutoff_input" -le 50 ]]; then
	word_size=2
else
	echo "please introduce a valid cluster percentage value between 0.4 and 1"
	exit 1
fi



if [ ! $output_dir ]; then
	output_dir=$(dirname $input_file)
	echo "Default output directory is" $output_dir
	mkdir -p $output_dir
else
	echo "Output directory is" $output_dir
	mkdir -p $output_dir
fi

if [ ! $file_name ]; then
	file_name=$(basename $input_file)
	echo "filename is" $file_name
fi

##CD-HIT EXECUTION

echo "$(date)"
echo "Clustering sequences with identity" $cluster_cutoff_input"% or higher"
echo "Using" $cd_hit_command "with file" $input_file
seq_number_prev_clstr=$(cat $input_file | grep ">" | wc -l)

cd $(dirname $input_file)

if [ -f $output_dir/$file_name""_""$cluster_cutoff_input ]; then \
	echo "Found a clustered file for sample" $file_name;
	echo "Omitting clustering process calculation"
	exit 1
else
	if [ $cd_hit_command  == "psi-cd-hit.pl" ]; then 
		$cd_hit_command -i $(basename $input_file) -o $file_name""_""$cluster_cutoff_input -c $cluster_cutoff -G $global_psi_cd_hit -g 1 -prog $psi_cd_hit_program -circle $is_circle
	
	else

		$cd_hit_command -i $(basename $input_file) -o $file_name""_""$cluster_cutoff_input -c $cluster_cutoff -n $word_size -d 0 -s $length_cutoff -B 1 -M $max_memory -T $threads
	fi
fi

seq_number_post_clstr=$(cat $file_name""_""$cluster_cutoff_input | grep ">" | wc -l)

echo "$(date)"
echo "DONE Clustering sequences with identity" $cluster_cutoff_input"% or higher"
echo "fasta file can be found in" $output_dir/$file_name""_""$cluster_cutoff_input
echo "Previous number of sequences=" $seq_number_prev_clstr
echo -e "Number of sequences after clustering=" $seq_number_post_clstr "\n"
cd $cwd
